src/chan_vese.jl

```diff
@@ -0,0 +1,252 @@
+using MosaicViews
+using LazyArrays
+using BenchmarkTools
+using Images, TestImages
+using ImageBase.ImageCore: GenericGrayImage, GenericImage
+
+function calculate_averages(img::AbstractArray{T, N}, H𝚽::AbstractArray{T, M}) where {T<:Real, N, M}
+    H𝚽ⁱ = @. 1. - H𝚽
+    ∫H𝚽 = sum(H𝚽)
+    ∫H𝚽ⁱ = sum(H𝚽ⁱ)
+    if ndims(img) == 2
+        ∫u₀H𝚽 = sum(img .* H𝚽)
+        ∫u₀H𝚽ⁱ = sum(img .* H𝚽ⁱ)
+    elseif ndims(img) == 3
+        ∫u₀H𝚽 = sum(img .* H𝚽, dims=(1, 2))
+        ∫u₀H𝚽ⁱ = sum(img .* H𝚽ⁱ, dims=(1, 2))
+    end
+    if ∫H𝚽 != 0
+        c₁ = ∫u₀H𝚽 / ∫H𝚽
+    end
+    if ∫H𝚽ⁱ != 0
+        c₂ = ∫u₀H𝚽ⁱ / ∫H𝚽ⁱ
+    end
+
+    return c₁, c₂
+end
+
+function difference_from_average_term(img::AbstractArray{T, N}, H𝚽::AbstractArray{T, M}, λ₁::Float64, λ₂::Float64) where {T<:Real, N, M}
+    c₁, c₂ = calculate_averages(img, H𝚽)
+
+    if ndims(img) == 2
+        return @. -λ₁ * (img - c₁)^2 + λ₂ * (img - c₂)^2
+    elseif ndims(img) == 3
+        return -λ₁ .* sum((img .- c₁).^2, dims=3) .+ λ₂ .* sum((img .- c₂).^2, dims=3)
+    end
+end
+# H𝚽 = LazyArray(@~ @. 1. * (𝚽ⁿ > 0))
+function _calculate_averages(img::AbstractArray{T, N}, 𝚽ⁿ::AbstractArray{T, M}) where {T<:Real, N, M}
+    ∫H𝚽 = ∫H𝚽ⁱ = ∫u₀H𝚽 = ∫u₀H𝚽ⁱ = 0
+
+    for i in CartesianIndices(img)
+        H𝚽 = 1. * (𝚽ⁿ[i] > 0)
+        H𝚽ⁱ = 1. - H𝚽
+        ∫H𝚽 += H𝚽
+        ∫H𝚽ⁱ += H𝚽ⁱ
+        ∫u₀H𝚽 += img[i] * H𝚽
+        ∫u₀H𝚽ⁱ += img[i] * H𝚽ⁱ
+    end
+    if ∫H𝚽 != 0
+        c₁ = ∫u₀H𝚽 ./ ∫H𝚽
+    end
+    if ∫H𝚽ⁱ != 0
+        c₂ = ∫u₀H𝚽ⁱ ./ ∫H𝚽ⁱ
+    end
+
+    return c₁, c₂
+end
+
+function δₕ(x::AbstractArray{T,N}, h::Float64=1.0) where {T<:Real, N}
+    return @~ @. h / (h^2 + x^2)
+end
+
+function initial_level_set(shape::Tuple)
```

Comment:
You can generalize this to arbitrary dimensions like this: `broadcast(*, kernelfactors((0:3, 0:5, 0:7))...)` where `kernelfactors` is from ImageFiltering.